File: 239-sliding-window-maximum/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& arr, int k) {
        deque<int>dq;
        vector<int>ans;
        int st=0,n=arr.size();

        for(int i=0;i<n;i++){
            while(!dq.empty() and dq.back()<arr[i]){
                dq.pop_back();
            }
            dq.push_back(arr[i]);
            
            if((i-st+1)==k){
                ans.push_back(dq.front());
                if(dq.front()==arr[st]){
                    dq.pop_front();
                }
                st++;
            }
        }
        return ans;
        
    }
};

File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        ans=[]
        dq=deque()
        st=0
        for i in range(len(nums)):
            while len(dq)!=0 and dq[-1]<nums[i]:
                dq.pop()
            dq.append(nums[i])
            if (i-st+1)==k:
                ans.append(dq[0])
                if dq[0]==nums[st]:
                    dq.popleft()
                st+=1   

        return ans                